intelliserver/api/functions/data/index.js

```diff
@@ -0,0 +1,105 @@
+const path = require('path');
+const axios = require('axios');
+const router = express.Router();
+const { ChatGPTInput, LLamaReplicateInput, LLamaSageInput, CohereInput } = require('intellinode');
+const { Chatbot, SupportedChatModels } = require('intellinode');
+const { USE_DEFAULT_KEYS } = require(path.join(global.__basedir, 'config'));
+
+const keys = {
+    'openai': process.env.OPENAI_API_KEY,
+    'replicate': process.env.REPLICATE_API_KEY,
+    'sagemaker': process.env.SAGEMAKER_API_KEY,
+    'cohere': process.env.COHERE_API_KEY
+};
+
+async function callSemanticSearchAPI(data) {
+    try {
+        const apiUrl = 'https://1cw9da6wl7.execute-api.us-east-2.amazonaws.com/semantic_search/';
+
+        const formData = new FormData();
+        Object.keys(data).forEach((key) => {
+            formData.append(key, data[key]);
+        });
+
+        const response = await axios.post(apiUrl, formData, {
+            headers: {
+                'Content-Type': 'multipart/form-data',
+            },
+        });
+
+        console.log(response.data);
+        return response.data;
+    } catch (error) {
+        throw new Error(`Error calling semantic search API: ${error.message}`);
+    }
+}
+
+function getChatbot(req) {
+    console.log('getChatbot provider: ', req.body.provider);
+    if (USE_DEFAULT_KEYS && !req.body.api_key) {
+        return new Chatbot(keys[req.body.provider.toLowerCase()], req.body.provider);
+    } else {
+        return new Chatbot(req.body.api_key, req.body.provider);
+    }
+}
+
+function addMessages(chatInput, messages) {
+    messages.forEach(msg => {
+        if (msg.role.toLowerCase() === "user") {
+            chatInput.addUserMessage(msg.content);
+        } else if (msg.role.toLowerCase() === "assistant") {
+            chatInput.addAssistantMessage(msg.content);
+        } else {
+            throw new Error(`Invalid message role: ${msg.role}`);
+        }
+    });
+    return chatInput;
+}
+
+function getChatInput(input, provider) {
+    const inputInst = input instanceof ChatGPTInput ? input :
+        provider === SupportedChatModels.REPLICATE ? new LLamaReplicateInput(input.system, input) :
+        provider === SupportedChatModels.SAGEMAKER ? new LLamaSageInput(input.system, input) :
+        provider === SupportedChatModels.COHERE ? new CohereInput(input.system, input) :
+            new ChatGPTInput(input.system, input);
+
+    return addMessages(inputInst, input.messages);
+}
+
+// New API route: data/chatbot
+router.post('/chatbot', async (req, res, next) => {
+    try {
+        // Call the semantic search API asynchronously
+        const semanticSearchResponse = await callSemanticSearchAPI(req.body);
+
+        // Concatenate the text from response data
+        const concatenatedText = semanticSearchResponse.data.map(document => {
+            return document.data.map(item => item.text).join(' ');
+        }).join(' ');
+        req.body.input.messages[0].content = concatenatedText;
```

Comment:
you should send the user query and the response of the semantic search to the chatbot as follow:

context: <semantic_search_response>

user query: <user_query>


best to add text file template in public or root, then load it and replace the place holders.

Reply:
feedback implemented